gRPC LB fixes from end two end testing

dgquintas opened the pull request:
Many fixes, which include:
- Handling of LB servers restarts.
- Backoff retries for LB server reconnection.
- Added missing weak references for `query_for_backends_locked()` calls.
- Wrong RR handover placement: moved from the RR connectivity callback to the LB call cb.
- Fixed data races with the addition of missing locks.
-  Fixed a leak that'd happen if the query to the LB server happened right before the shutdown of the LB.

dgquintas commented:
@slash-lib FYI.

markdroth commented on src/core/ext/lb_policy/grpclb/grpclb.c:
```diff
@@ -296,20 +300,53 @@ typedef struct glb_lb_policy {
+  /*  client data associated with the LB server communication */
+  /************************************************************/
+  /* called once initial metadata's been sent */
+  grpc_closure md_sent;
```
Doesn't look like this is used anywhere.

markdroth commented on src/core/ext/lb_policy/grpclb/grpclb.c:
```diff
@@ -916,250 +950,181 @@ static void glb_notify_on_state_change(grpc_exec_ctx *exec_ctx,
-
-static lb_client_data *lb_client_data_create(glb_lb_policy *glb_policy) {
+static void res_recv_cb(grpc_exec_ctx *exec_ctx, void *arg, grpc_error *error);
+static void lb_client_init(glb_lb_policy *glb_policy) {
```
This isn't actually initializing the LB client (i.e., it doesn't initialize the LB channel), just an individual call.  So I suggest calling it `lb_call_init()`.

markdroth commented on src/core/ext/lb_policy/grpclb/grpclb.c:
```diff
@@ -916,250 +950,181 @@ static void glb_notify_on_state_change(grpc_exec_ctx *exec_ctx,
-  grpc_call_destroy(lb_client->lb_call);
-  grpc_metadata_array_destroy(&lb_client->initial_metadata_recv);
-  grpc_metadata_array_destroy(&lb_client->trailing_metadata_recv);
+static void lb_client_destroy(glb_lb_policy *glb_policy) {
```
Similar comment here: Suggest calling this `lb_call_destroy()`.

markdroth commented on src/core/ext/lb_policy/grpclb/grpclb.c:
```diff
@@ -296,20 +300,53 @@ typedef struct glb_lb_policy {
+  grpc_closure req_sent;
+
+  /* A response from the LB server has been received (or error). Process it */
+  grpc_closure res_rcvd;
```
A few naming suggestions:

- Please prefix all of the fields that were formerly in the `lb_client_data` struct with `lb_`, so that it's clear which ones are related to communication with the LB.
- Please avoid abbreviations (e.g., prefer `lb_response_received()` to `res_rcvd()`).
- I suggest using the same name for each closure and the corresponding callback function.

markdroth commented on src/core/ext/lb_policy/grpclb/grpclb.c:
```diff
@@ -296,20 +300,53 @@ typedef struct glb_lb_policy {
+
+  /* called once the LoadBalanceRequest has been sent to the LB server. See
+   * src/proto/grpc/.../load_balancer.proto */
+  grpc_closure req_sent;
```
This doesn't appear to be used anywhere either.

markdroth commented on src/core/ext/lb_policy/grpclb/grpclb.c:
```diff
@@ -916,250 +950,181 @@ static void glb_notify_on_state_change(grpc_exec_ctx *exec_ctx,
       if (grpc_lb_glb_trace) {
         gpr_log(GPR_INFO, "Serverlist with %lu servers received",
                 (unsigned long)serverlist->num_servers);
+        /* TODO(dgq): this needs to work with ipv6. */
```
Can we re-use the code from `process_serverlist()` for this, so that we don't have to maintain similar code twice?

dgquintas replied:
Done. Added `parse_server`

markdroth commented on src/core/ext/lb_policy/grpclb/grpclb.c:
```diff
@@ -916,250 +950,181 @@ static void glb_notify_on_state_change(grpc_exec_ctx *exec_ctx,
         gpr_log(GPR_INFO, "Serverlist with %lu servers received",
                 (unsigned long)serverlist->num_servers);
+        /* TODO(dgq): this needs to work with ipv6. */
+        for (size_t i = 0; i < serverlist->num_servers; ++i) {
```
Do we really want to do this unconditionally?  Or would it make sense to wrap this loop in a `if (grpc_lb_glb_trace) {...}`?

markdroth commented on src/core/ext/lb_policy/grpclb/grpclb.c:
```diff
@@ -1185,42 +1150,105 @@ static void res_recv_cb(grpc_exec_ctx *exec_ctx, void *arg, grpc_error *error) {
-        exec_ctx, lb_client->lb_call, ops, (size_t)(op - ops),
-        &lb_client->close_sent);
-    GPR_ASSERT(GRPC_CALL_OK == call_error);
+    grpc_call_cancel(glb_policy->lb_call, NULL);
```
Do we really want to cancel instead of sending a close from the client?  It seems like that might make the logs on the LB side look like every call is cancelled instead of being completed successfully, which might mask problems in the case where there really is a cancellation.

dgquintas replied:
Actually, I don't think we want to cancel/close. That'd only result in a retry and reconnection to the same channel, which, by the pick first policy, would likely end up at the same server. I've changed to code to log and ignore. Do you think more drastic measures are necessary? Such as eliminating the server that returned the invalid serverlist and recreating the channel.

markdroth commented on src/core/ext/lb_policy/grpclb/grpclb.c:
```diff
@@ -1185,42 +1150,105 @@ static void res_recv_cb(grpc_exec_ctx *exec_ctx, void *arg, grpc_error *error) {
-  /* empty payload: call cancelled by server. Cleanups happening in
-   * srv_status_rcvd_cb */
+  /* else, empty payload: call cancelled by server. */
+  grpc_metadata_array_destroy(&glb_policy->initial_metadata_recv);
```
Why does this need to happen here, instead of in `srv_status_rcvd_cb()`?

markdroth commented on src/core/ext/lb_policy/grpclb/grpclb.c:
```diff
@@ -916,250 +950,181 @@ static void glb_notify_on_state_change(grpc_exec_ctx *exec_ctx,
+static void query_for_backends_locked(grpc_exec_ctx *exec_ctx,
+                                      glb_lb_policy *glb_policy) {
   GPR_ASSERT(glb_policy->lb_channel != NULL);
+  GRPC_LB_POLICY_WEAK_REF(&glb_policy->base, "query_for_backends_locked");
```
How do the semantics of weak refs differ from normal refs, and why are we using them here?

markdroth replied:
It's probably worth adding a comment indicating that this ref will eventually be released by `srv_status_rcvd_cb()`.

dgquintas replied:
Weak refs don't prevent the shutdown of the LB policy. When no strong refs are left but there are still weak ones, shutdown is called. When both strong and weak refs are zero, destroy is called. In this case we need a weak one because the existence of the call to the LB server shouldn't preclude a call to `glb_shutdown` (where the call will be terminated if ongoing).

If we used a strong ref, we'd have to wait for the eventual timeout of the LB call for the corresponding strong unref to happen as part of the cleanup performed by `srv_status_rcvd_cb` (pre rename).

dgquintas replied:
done

markdroth replied:
I don't see lb_policy.h in this PR.

markdroth commented on src/core/ext/lb_policy/grpclb/grpclb.c:
```diff
@@ -1185,42 +1150,105 @@ static void res_recv_cb(grpc_exec_ctx *exec_ctx, void *arg, grpc_error *error) {
+            (void *)glb_policy->lb_call);
+  }
+
+  if (glb_policy->lb_call_status == GRPC_STATUS_UNIMPLEMENTED) {
```
I thought that we decided that we were not going to treat UNIMPLEMENTED specially -- we should do the same thing for any non-OK status.

markdroth commented on src/core/ext/lb_policy/grpclb/grpclb.c:
```diff
@@ -1185,42 +1150,105 @@ static void res_recv_cb(grpc_exec_ctx *exec_ctx, void *arg, grpc_error *error) {
+    while ((pp = glb_policy->pending_picks)) {
+      glb_policy->pending_picks = pp->next;
+      if (grpc_lb_glb_trace) {
+        gpr_log(GPR_INFO, "Cancelling pending pick: %s", error_desc);
```
For requests with wait_for_ready=false (the default), do we actually want to cancel the picks?  Shouldn't the picks stay pending until either we successfully retry the LB request or the calls' deadlines expire?  Or is the pick retries for us by the client_channel code?

markdroth commented on src/core/ext/lb_policy/round_robin/round_robin.c:
```diff
@@ -269,7 +269,9 @@ static void rr_destroy(grpc_exec_ctx *exec_ctx, grpc_lb_policy *pol) {
     subchannel_data *sd = p->subchannels[i];
     GRPC_SUBCHANNEL_UNREF(exec_ctx, sd->subchannel, "round_robin_destroy");
-    sd->user_data_vtable.destroy(sd->user_data);
+    if (sd->user_data_vtable != NULL) {
```
I would check `sd->user_data` instead of `sd->user_data_vtable`.  The vtable is guaranteed to be non-NULL if the user_data is non-NULL (see https://github.com/grpc/grpc/blob/master/src/core/ext/client_channel/lb_policy_factory.c#L75), but what we really care about here is whether the user_data is non-NULL.

dgquintas replied:
Done. Added assert for documentation purposes.

markdroth commented on src/core/ext/lb_policy/grpclb/grpclb.c:
```diff
@@ -1007,8 +1011,10 @@ static void lb_client_destroy(glb_lb_policy *glb_policy) {
   GPR_ASSERT(glb_policy->lb_channel != NULL);
+  /* take a weak ref (won't prevent calling of \a glb_shutdown if the strong ref
+   * count goes to zero) to be unref'd in lb_on_server_status_received */
   GRPC_LB_POLICY_WEAK_REF(&glb_policy->base, "query_for_backends_locked");
```
Now that I think about it, don't we actually need to take two weak refs here, one for each of the pending callbacks we're about to create (`lb_on_server_status_received` and `lb_on_response_received`)?  If we get shut down, they will probably both be in flight at the same time, and we'll need to handle both of them.

dgquintas replied:
Yes, good point. Done.

markdroth replied:
I don't see any change here.

markdroth commented on src/core/ext/lb_policy/grpclb/grpclb.c:
```diff
@@ -1132,29 +1133,25 @@ static void res_recv_cb(grpc_exec_ctx *exec_ctx, void *arg, grpc_error *error) {
-    gpr_slice_unref(response_slice);
-    grpc_call_cancel(glb_policy->lb_call, NULL);
-    /* srv_status_rcvd_cb will pick up the cancellation and clean up */
+    /* keep listening for serverlist updates */
```
I think we need to avoid doing this if `glb_policy->shutting_down` is true.  Otherwise, we will perpetually have this callback either executing or pending -- it will never actually go away.

dgquintas replied:
I've added it for clarity, but it's not strictly necessary: `glb_shutdown` terminates the call, which means all pending callbacks will be called, and in this particular case, `glb_policy->lb_response_payload` would be NULL, which would make it go out the way of the `else, empty payload: call cancelled.` comment (which I've also changed slightly).

dgquintas commented on src/core/ext/lb_policy/grpclb/grpclb.c:
```diff
@@ -47,15 +47,12 @@
+ * lb_on_response_received. The former will be called when the call to the LB
+ * server completes. This can happen if the LB server closes the connection or
+ * if this policy itself cancels the call (for example because it's shutting
+ * down).If the internal call times out, the usual behavior of pick-first
```
Done.

dgquintas commented on src/core/ext/lb_policy/grpclb/grpclb.c:
```diff
@@ -311,32 +308,28 @@ typedef struct glb_lb_policy {
+  grpc_byte_buffer *lb_response_payload;
 
-  /* the call's status and status detailset in srv_status_rcvd_cb() */
+  /* the call's status and status detailset in lb_on_server_status_received() */
```
Done

dgquintas commented on src/core/ext/lb_policy/grpclb/grpclb.c:
```diff
@@ -1191,40 +1188,11 @@ static void srv_status_rcvd_cb(grpc_exec_ctx *exec_ctx, void *arg,
+  lb_call_destroy(glb_policy);
 
   if (!glb_policy->shutting_down) {
     GPR_ASSERT(!was_cancelled);
```
You are right. Removed assert. It always felt overly restrictive anyway.

dgquintas commented:
Addressing your comments in an ad-hoc way because github:

> The only substantive remaining change is the reference that needs to be taken for the > lb_on_server_status_received() callback.

I don't know which one you are referring to. `lb_on_server_status_received` takes a weak ref on line 1218. ("grpclb_retry_timer").

> Removing the assertion is good, but my broader question here still stands: If the LB server shuts down, shouldn't we go back and pick a new one? Or will the client continue to try to use the bad server forever?
If the server shuts down, the `lb_on_server_status_received` would be called with `glb_policy->shutting_down` being false, which would result in a retry. Because the lb_call is recreated, the pick first policy in the lb_channel should pick a healthy LB server if one is available.
